pkg/ingester/ingester_v2.go

```diff
@@ -1280,3 +1312,435 @@ func (i *Ingester) v2FlushHandler(w http.ResponseWriter, _ *http.Request) {
 
 	w.WriteHeader(http.StatusNoContent)
 }
+
+// createBackfillTSDB creates a TSDB for a given userID and bucketName, and returns the created db.
+func (i *Ingester) createBackfillTSDB(userID, bucketName string, blockRange int64) (*userTSDB, error) {
+	tsdbPromReg := prometheus.NewRegistry()
+	udir := i.cfg.TSDBConfig.BackfillBlocksDir(userID)
+	dbDir := filepath.Join(udir, bucketName)
+	userLogger := util.WithUserID(userID, util.Logger)
+
+	userDB := &userTSDB{
+		userID:              userID,
+		refCache:            cortex_tsdb.NewRefCache(),
+		ingestedAPISamples:  newEWMARate(0.2, i.cfg.RateUpdatePeriod),
+		ingestedRuleSamples: newEWMARate(0.2, i.cfg.RateUpdatePeriod),
+		lastUpdate:          atomic.NewInt64(0),
+	}
+
+	// Create a new user database
+	db, err := tsdb.Open(dbDir, userLogger, tsdbPromReg, &tsdb.Options{
+		RetentionDuration: i.cfg.TSDBConfig.Retention.Milliseconds(),
+		MinBlockDuration:  blockRange,
+		MaxBlockDuration:  blockRange,
+		NoLockfile:        true,
+		StripeSize:        i.cfg.TSDBConfig.StripeSize,
+		WALCompression:    i.cfg.TSDBConfig.WALCompressionEnabled,
+	})
+	if err != nil {
+		return nil, err
+	}
+	db.DisableCompactions() // we will compact on our own schedule
+
+	userDB.DB = db
+	userDB.setLastUpdate(time.Now()) // After WAL replay.
+
+	// Thanos shipper requires at least 1 external label to be set. For this reason,
+	// we set the tenant ID as external label and we'll filter it out when reading
+	// the series from the storage.
+	l := labels.Labels{
+		{
+			Name:  cortex_tsdb.TenantIDExternalLabel,
+			Value: userID,
+		}, {
+			Name:  cortex_tsdb.IngesterIDExternalLabel,
+			Value: i.lifecycler.ID,
+		},
+	}
+
+	// Create a new shipper for this database
+	if i.cfg.TSDBConfig.ShipInterval > 0 {
+		userDB.shipper = shipper.New(
+			userLogger,
+			tsdbPromReg,
+			udir,
+			cortex_tsdb.NewUserBucketClient(userID, i.TSDBState.bucket),
+			func() labels.Labels { return l },
+			metadata.ReceiveSource,
+			true, // Allow out of order uploads. It's fine in Cortex's context.
+		)
+	}
+
+	//i.TSDBState.tsdbMetrics.setRegistryForUser(userID, tsdbPromReg)
+	return userDB, nil
+}
+
+func (i *Ingester) v2BackfillPush(userID string, la []client.LabelAdapter, s client.Sample) error {
+	bucket, err := i.getOrCreateBackfillTSDB(userID, s.TimestampMs)
```

Comment:
We shouldn't call `getOrCreateBackfillTSDB` for every single sample to append. As mentioned in another comment, I think would be better if we:
1. Get a list of all samples to backfill
2. Partition these samples into buckets
3. We ingest each bucket, 1 by 1, and for each bucket we "get or create" the backfill TSDB